File: web/src/engine/predictions/cyclingPower.ts
// Portage de Predictions/CyclingPower.swift — modèle physique puissance ⇄ vitesse.
import type { BikeType } from '../domain/types';

export interface CyclingPowerParams {
  totalMassKg: number;
  cda: number; // coefficient de traînée × surface (m²)
  crr?: number; // résistance au roulement
  airDensity?: number; // ρ (kg/m³)
  drivetrainEfficiency?: number;
  gravity?: number;
}

export class CyclingPowerModel {
  totalMassKg: number;
  cda: number;
  crr: number;
  airDensity: number;
  drivetrainEfficiency: number;
  gravity: number;

  constructor(p: CyclingPowerParams) {
    this.totalMassKg = p.totalMassKg;
    this.cda = p.cda;
    this.crr = p.crr ?? 0.005;
    this.airDensity = p.airDensity ?? 1.225;
    this.drivetrainEfficiency = p.drivetrainEfficiency ?? 0.98;
    this.gravity = p.gravity ?? 9.81;
  }

  /** CdA typique selon le type de vélo / la position aéro. */
  static typicalCdA(bikeType: BikeType, aeroBars: boolean): number {
    switch (bikeType) {
      case 'tt': return aeroBars ? 0.24 : 0.28;
      case 'road': return aeroBars ? 0.29 : 0.32;
      case 'gravel': return 0.36;
      case 'mtb': return 0.42;
      case 'trainer': return 0.32;
    }
  }

  /** Puissance pédale (W) requise pour une vitesse (m/s) et une pente. */
  power(v: number, grade = 0, headwind = 0): number {
    const rolling = this.totalMassKg * this.gravity * this.crr;
    const climbing = this.totalMassKg * this.gravity * grade;
    const apparent = v + headwind;
    const aero = 0.5 * this.airDensity * this.cda * apparent * apparent;
    const pWheel = (rolling + climbing) * v + aero * v;
    return pWheel / this.drivetrainEfficiency;
  }

  /** Vitesse (m/s) atteinte pour une puissance (W) donnée — inversion numérique. */
  speed(targetPower: number, grade = 0, headwind = 0): number {
    if (!(targetPower > 0)) throw new Error('speed: puissance > 0');
    let lo = 0.0;
    let hi = 30.0; // jusqu'à 108 km/h
    for (let i = 0; i < 100; i++) {
      const mid = (lo + hi) / 2;
      if (this.power(mid, grade, headwind) < targetPower) lo = mid;
      else hi = mid;
      if (hi - lo < 1e-4) break;
    }
    return (lo + hi) / 2;
  }

  /** Temps (s) prédit sur une distance (m) à une puissance soutenue donnée. */
  predictTimeSeconds(distanceM: number, sustainedPowerW: number, grade = 0): number {
    const v = this.speed(sustainedPowerW, grade);
    return distanceM / v;
  }
}

File: Packages/TriathlonEngine/Sources/TriathlonEngine/Predictions/CyclingPower.swift
import Foundation

/// Modèle physique puissance ⇄ vitesse en cyclisme.
///
/// `P_roue = (m·g·Crr + m·g·pente)·v + ½·ρ·CdA·(v+vent)²·v`
/// `P_pédale = P_roue / rendement`
public struct CyclingPowerModel: Sendable, Equatable {
    public var totalMassKg: Double        // cycliste + vélo + équipement
    public var cda: Double                // coefficient de traînée × surface (m²)
    public var crr: Double                // résistance au roulement
    public var airDensity: Double         // ρ (kg/m³)
    public var drivetrainEfficiency: Double
    public var gravity: Double

    public init(
        totalMassKg: Double,
        cda: Double,
        crr: Double = 0.005,
        airDensity: Double = 1.225,
        drivetrainEfficiency: Double = 0.98,
        gravity: Double = 9.81
    ) {
        self.totalMassKg = totalMassKg
        self.cda = cda
        self.crr = crr
        self.airDensity = airDensity
        self.drivetrainEfficiency = drivetrainEfficiency
        self.gravity = gravity
    }

    /// CdA typique selon le type de vélo / la position aéro.
    public static func typicalCdA(bikeType: BikeType, aeroBars: Bool) -> Double {
        switch bikeType {
        case .tt: return aeroBars ? 0.24 : 0.28
        case .road: return aeroBars ? 0.29 : 0.32
        case .gravel: return 0.36
        case .mtb: return 0.42
        case .trainer: return 0.32
        }
    }

    /// Puissance pédale (W) requise pour une vitesse (m/s) et une pente (fraction, ex. 0.03).
    public func power(forSpeed v: Double, grade: Double = 0, headwind: Double = 0) -> Double {
        let rolling = totalMassKg * gravity * crr
        let climbing = totalMassKg * gravity * grade
        let apparent = v + headwind
        let aero = 0.5 * airDensity * cda * apparent * apparent
        let pWheel = (rolling + climbing) * v + aero * v
        return pWheel / drivetrainEfficiency
    }

    /// Vitesse (m/s) atteinte pour une puissance (W) donnée — inversion numérique.
    public func speed(forPower target: Double, grade: Double = 0, headwind: Double = 0) -> Double {
        precondition(target > 0)
        var lo = 0.0, hi = 30.0            // jusqu'à 108 km/h
        for _ in 0..<100 {
            let mid = (lo + hi) / 2
            if power(forSpeed: mid, grade: grade, headwind: headwind) < target { lo = mid } else { hi = mid }
            if hi - lo < 1e-4 { break }
        }
        return (lo + hi) / 2
    }

    /// Temps (s) prédit sur une distance (m) à une puissance soutenue donnée.
    public func predictTimeSeconds(distanceM d: Double, sustainedPowerW p: Double, grade: Double = 0) -> Double {
        let v = speed(forPower: p, grade: grade)
        return d / v
    }
}
